crates/chain/src/persist.rs

```diff
@@ -1,33 +1,39 @@
-use core::convert::Infallible;
-
 use crate::Append;
+use alloc::boxed::Box;
+use anyhow::Result;
+use core::fmt;
 
 /// `Persist` wraps a [`PersistBackend`] (`B`) to create a convenient staging area for changes (`C`)
 /// before they are persisted.
 ///
 /// Not all changes to the in-memory representation needs to be written to disk right away, so
 /// [`Persist::stage`] can be used to *stage* changes first and then [`Persist::commit`] can be used
 /// to write changes to disk.
-#[derive(Debug)]
-pub struct Persist<B, C> {
-    backend: B,
+pub struct Persist<C> {
+    backend: Box<dyn PersistBackend<C>>,
     stage: C,
 }
 
-impl<B, C> Persist<B, C>
+impl<C: fmt::Debug> fmt::Debug for Persist<C> {
+    fn fmt(&self, fmt: &mut core::fmt::Formatter<'_>) -> Result<(), core::fmt::Error> {
+        write!(fmt, "{:?}", self.stage)?;
+        Ok(())
+    }
+}
+
+impl<C> Persist<C>
 where
-    B: PersistBackend<C>,
     C: Default + Append,
 {
     /// Create a new [`Persist`] from [`PersistBackend`].
-    pub fn new(backend: B) -> Self {
+    pub fn new(backend: Box<dyn PersistBackend<C>>) -> Self {
```

Comment:
I think it is nicer to hide the `Box` from the caller here. I can't think of a benefit for exposing it?

```suggestion
    pub fn new(backend: impl PersistBackend<C> + 'static) -> Self {
```